### DISCUSSION TITLE
Cannot stringify arbitrary non-POJOs

### DISCUSSION THREAD

**Discussion by wvhulle** [sveltejs/kit#7259] [Q&A] [answered]
When adding a +page.server.ts file and filling it with the contents

```
import { json } from "@sveltejs/kit";

export async function load({ params }) {
    return json()
}
```

SvelteKit throws the error 

```
DevalueError: Cannot stringify arbitrary non-POJOs
    at walk (file:///home/<project>/node_modules/.pnpm/devalue@4.0.0/node_modules/devalue/src/uneval.js:83:13)
```

Is this a regression in SvelteKit caused by devalue bumping to 4.0.0?

**Comment by geoffrich** [comment]
This is expected. You [only return POJOs](https://kit.svelte.dev/docs/load) (plain-old-javascript-objects) from `load` functions, e.g.

```js
export async function load({ params }) {
    return { hello: 'world', count: 2 }
}
```

The `json` helper creates a `Response`, which you can't return from `load`. Only `+server.js` files [return Responses](https://kit.svelte.dev/docs/routing#server).

**Reply by wvhulle** [reply]
Thank you, yes I figured it out. Maybe the error could be clearer.

**Reply by Rich-Harris** [reply]
Opened an issue https://github.com/sveltejs/kit/issues/7262

**Comment by RossAddinall** [comment]
Just a quick note on this.  I'm new to Svelte coming from React.
It seems odd to me that you can use the "fail" helper in a +page.server.js but not the "json" helper.
See [this](https://kit.svelte.dev/docs/form-actions#anatomy-of-an-action-validation-errors) example.

**Reply by geoffrich** [reply]
What are you trying to do that requires returning `json(...)` in a `+page.server.js`? `json` is only intended for use in `+server.js` endpoints that require you to return a [Response object](https://developer.mozilla.org/en-US/docs/Web/API/Response). There is no need for it in `+page.server.js`

In `+page.server.js` you just return data objects directly, e.g.

```js
// ✅ do this!
export function load() {
    return { hello: 'world' };
}

// ❌ don't do this!
export function load() {
    return json({ hello: 'world' });
}
```

**Comment by RossAddinall** [comment]
@geoffrich Yes, I realise that now.  I was trying to be smart and saw that the example used "fail" and in my enthusiasm thought I would use "json" to return the successful result.  Now that I understand that "json" forms a Response object I can see that is not what I want to do, but initially I saw "fail", "json" etc. as a set of helpers and naively if I could use one I could use them all.

Thanks for your clarification, hopefully, this might help someone else.

**Comment by ridly** [comment]
Hello guys,

I know I'm late for the party. I just ran into this problem myself when using Prisma.Decimal field type in the database. I found that superjson can `serialize` / `deserialize` it, including other complex types like `Date`, `BigInt`.
I'm wondering if we can replace sveltekit's `JSON.stringify` / `JSON.parse` with superjson? (https://github.com/flightcontrolhq/superjson)
Let me know.

Cheers
Tamas

**Reply by Rich-Harris** [reply]
https://github.com/sveltejs/kit/discussions/7259#discussioncomment-13210638

**Comment by Rich-Harris** [comment, marked as answer]
You can now use the [transport hook](https://svelte.dev/docs/kit/hooks#Universal-hooks-transport) to define custom encoders and decoders.

**Reply by ridly** [reply]
Thanks @Rich-Harris , I'll give it a go. You have a solution for everything.
